docs/proposals/ssh-known_hosts.rst

```diff
@@ -0,0 +1,211 @@
+.. vim:set ft=rst spell:
+
+Managing SSH Known Hosts
+========================
+
+This proposal suggests a method to more securely manage and ensure
+authentication of remote hosts through the SSH protocol within Tarmak
+environments on AWS.
+
+Background
+----------
+
+Currently we solely use the external OpenSHH SSH command line tool to connect to
+remote instances on EC2 for both interactive shell sessions as well as
+tunnelling and proxy commands for other services, including, connecting to the
+Vault cluster and the private Kubernetes API server endpoint. Currently in
+development is the replacement of our programmatic use cases of SSH in favour of
+the in package Go solution, a choice stemming from pain points developing more
+sophisticated utility functions for Tarmak and the desire for improvements in
+control of connections to remote hosts.
+
+During development of this replacement it became clear that proper care must be
+taken during authentication of host public keys during connection and manual
+management of our ``ssh_known_hosts`` cluster file. Our current implementation
+allows OpenSSH to maintain this file however, does not exit with an error if
+public keys do not match due to the flag ``StrictHostKeyChecking`` set to
+``no``. Not only does a miss-match in public keys not cause an error, the
+population of known public keys on different authenticated machines to the same
+EC2 hosts will always use the hosts presented public key, meaning the set of
+public keys could potentially be different for users accessing the same cluster.
+
+Objective
+---------
+
+By implementing stricter enforcement of the ``ssh_known_hosts`` file and passing
+it's management to Tarmak, we can improve the security of SSH connections to
+remote hosts. The key high level points to achieving this is as follows:
+
+ - Disable writes from the OpenSSH command to the ``ssh_known_hosts`` file and
+   enforce strict checking.
+ - Enforce that our in package implementation of SSH connections adheres to this
+   file also.
+ - Collect public keys during instance start up that are then stored, tightly
+   coupled with that host. These keys are able to be used as a source of truth
+   for other authenticated users attempting to connect to remote hosts on the
+   cluster that have empty or an incomplete ``ssh_known_hosts`` file.
+
+Changes
+-------
+
+Firstly, we must restrict the OpenSSH command line tool from editing the
+``ssh_known_hosts`` file and strictly enforce it by updating the generator for the
+``ssh_config`` file. This enables Tarmak to take control of the ``ssh_known_hosts``
+file management.
+
+In order to create a source of truth for each host's public key, each instance
+will have it's public key's attached as a tags, shortly after boot time like the
```

Comment:
Also some considerations regarding if we can make sure terraform is not removing those tags every time we run it.

This is esp important for aws_instance resources. Maybe https://www.terraform.io/docs/configuration/resources.html#ignore_changes is good enough